dask/dataframe/tests/test_dataframe.py

```diff
@@ -2727,3 +2727,17 @@ def test_better_errors_object_reductions():
     with pytest.raises(ValueError) as err:
         ds.mean()
     assert str(err.value) == "`mean` not supported with object series"
+
+def test_sample_from_query():
+    def _mk_df(key, n_row = 100, n_col = 10):
+        ret = pd.DataFrame(np.zeros((n_row, n_col)), columns = ['c_%d' % j for j in range(n_col) ])
+        ret['key'] = key
+        return ret
+    ddf = dd.from_delayed([ dask.delayed(_mk_df)(k) for k in range(100) ])
+    ddf = ddf.repartition(npartitions=4)
+    assert ddf.npartitions == 4
+
+    # most of the partitions don't have key > 90
+    ddf1 = ddf.query("key > 90").sample(frac=0.1)
+    df = ddf1.compute()
+    assert np.all(df['key'] > 90)
```

Comment:
A simpler test (with a title more fitting of the issue):

```python
def test_sample_empty_partitions():
    @dask.delayed
    def make_df(n):
        return pd.DataFrame(np.zeros((n, 4)), columns=list('abcd'))
    ddf = dd.from_delayed([make_df(0), make_df(100), make_df(0)])
    ddf2 = ddf.sample(frac=0.2)
    # smoke test sample on empty partitions
    res = ddf2.compute()
    assert res.dtypes.equals(ddf2.dtypes)
```

Reply:
Can we simplify further?

```python
In [16]: df = pd.DataFrame({"A": [True, True, False, False]})

In [17]: ddf = dd.from_pandas(df, 2)

In [18]: ddf[ddf.A].sample(frac=.5).compute()
```

I haven't tested that, but I think it hits the same thing?

Reply:
I'd say that this is less clear that you have empty partitions, but sure, that tests the same thing.

Reply:
I think jcrist's simplified test is good. Do I need to anything for this to be merged? Thanks.

Reply:
> Do I need to anything for this to be merged?

Can you update the test as I described above?

Reply:
@xwang777 are you able to update this PR with the change suggested above?  If so then we can go ahead and merge this.  (thanks for your work!)